### PR TITLE
CLN: contextmanagers should yield in a try-finally block

### PR DESCRIPTION
**Pull request by twoertwein**
`@contextmanagers` should yield within a try-finally block to release/reset even when an error occurs.

### CODE REVIEW

**Comment by jreback** [comment]
what motivated this change? e.g. did you see an error on something?

**Comment by twoertwein** [comment]
> what motivated this change? e.g. did you see an error on something?

No error, was just grepping for `@contextmanager` to see whether they all use try-finally. I think at least the change in `file_leak_context` should be very useful to verify that IO even closes file handles in case of an error (not sure how important the other two changes are).

File: pandas/_config/config.py [hunk 1]
```diff
@@ -763,10 +763,12 @@ def inner(key: str, *args, **kwds):
     set_option = wrap(set_option)
     get_option = wrap(get_option)
     register_option = wrap(register_option)
-    yield None
-    set_option = _set_option
-    get_option = _get_option
-    register_option = _register_option
+    try:
+        yield None
```
Comment (jbrockmendel) [inline pandas/_config/config.py:767]: nitpick can this just be "yield" like the others

Comment (twoertwein) [reply pandas/_config/config.py]: done

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
thanks @twoertwein

**Pull request merged**